### PR TITLE
Separate user and dev doxygen

### PR DESCRIPTION
**Pull request by chrisdembia**
Part of #383. 
With this PR, doxygen is built twice. One build is intended for users (scripting users but also C++ users who are not making new classes), and developers (of plugins, contributing to opensim-core).

**I've uploaded both user and dev versions so you can explore them**. Click the images below:

1. [user](http://stanford.edu/~dembia/opensim-better-property-docs-20150522-user/):
[screenshot from 2015-05-22 18 13 22](http://stanford.edu/~dembia/opensim-better-property-docs-20150522-user/)

2. [developer](http://stanford.edu/~dembia/opensim-better-property-docs-20150522-developer/):
[screenshot from 2015-05-22 18 12 31](http://stanford.edu/~dembia/opensim-better-property-docs-20150522-developer/)

Here's what I did:

1. I created new doxyfiles that share most of the properties from a "shared" doxyfile. This is how the two separate builds are configured.
2. The two builds also use two different doxygen-layout*.xml files; in the user docs, the doxygen-layout file is edited so that the protected methods do not appear. James spent a lot of time making this doxygen-layout file so that the result is not cluttered and is useful or scripting users.
3. The conditional `\if user` and `\if developer`, as well as the conditional `\internal` doxygen syntax allows doxygen documentation writers to tailor their documentation. I describe this in CONTRIBUTING.md.
4. I added blurbs to OpenSimDoxygenMain.h to describe the two different types of doxygen.
5. user docs get installed into `sdk/doc/html_user` and dev docs into `sdk/doc/html_developer`. The OpenSimAPI.html file now links to the user docs.
6. I also moved the documentation folder from `/OpenSim/doc` to `/doc`, which is something that I am very excited about.

After this gets merged, there are a few other things to consider:

1. After all the documentation goes online, can we link the user docs to the dev docs, etc. so users can jump between the two?
2. We need to update the online links to doxygen  (from opensim.stanford.edu and from confluence) so there are links to the separate user and dev pages.

Please check out the changes!

/cc @jimmyDunne @jenhicks @aseth1 @aymanhab @klshrinidhi @sherm1

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -150,6 +150,11 @@ If the “copy stuff” part consists only of assignments that work for self ass
 
 Doxygen only looks in your .h files; it does not generate documentation from .cpp files. Thus, comments in .cpp files don't need to follow doxygen formatting, and in fact they should not because it is confusing and makes it look like there is API documentation when there isn't. You should mostly use `//`-style comments in .cpp files, and be sure you are addressing your comments to the right audience -- no doxygen reader will ever see them.
 
+We produce two doxygen websites, one for (scripting) users and one for C++
```
Comment (sherm1) [inline CONTRIBUTING.md]: "website" doesn't seem like the right word here. How about "We produce two sets of doxygen-generated API documentation" or something like that? Whether it gets posted to a web site is a separate issue.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -150,6 +150,11 @@ If the “copy stuff” part consists only of assignments that work for self ass
...
+developers. When writing doxygen comments, you can tailor comments to either
+users or developers using the `\if user ... \endif` or `\if developer ...
+\endif` syntax.
```
Comment (sherm1) [inline CONTRIBUTING.md]: If there are differences that are generated automatically (such as suppressing `protected` members) it is worth briefly describing those differences here I think.

File: OpenSim/OpenSimDoxygenMain.h [hunk 3]
```diff
@@ -1,5 +1,5 @@
-#ifndef _OpenSim_Doxygen_Main_h_
-#define _OpenSim_Doxygen_Main_h_
+#ifndef OPENSIM_DOXYGEN_MAIN_H_
+#define OPENSIM_DOXYGEN_MAIN_H_
```
Comment (sherm1) [inline OpenSim/OpenSimDoxygenMain.h]: Yay! Conforming header guard.

File: OpenSim/OpenSimDoxygenMain.h [hunk 4]
```diff
@@ -30,10 +30,30 @@ OpenSim source and it is not installed with OpenSim. **/
...
+site contains documentation for protected methods that you might need to
+implement. If you simply want to make use of the API in MATLAB or Python, then
+check out the \a user API documentation.
+\endif
+
 \htmlonly
 <!-- ImageReady Slices -->
-The table below represents the conceptual heirarchy of OpenSim within SimTK. Each box represents a class utilized by the OpenSim API, and is built on top of (i.e., requires) the components underneath. 
-For example, you can click on the "Model Component" block to see a list of model components, such as body, constraint, or joint.
+The table below represents the conceptual heirarchy of OpenSim within SimTK.
```
Comment (sherm1) [inline OpenSim/OpenSimDoxygenMain.h]: This use of "SimTK" is outdated. Something like: "The table below represents the hierarchical structure of OpenSim."

File: OpenSim/OpenSimDoxygenMain.h [hunk 5]
```diff
@@ -30,10 +30,30 @@ OpenSim source and it is not installed with OpenSim. **/
 
 /** @mainpage  OpenSim 3.2 Documentation
 
+\if user
+This documentation is for people who want to use the OpenSim application
+programming interface (API) with the MATLAB, Python, Java or C++ languages.
+Therefore, internal (public/private) methods are hidden from this
+documentation. If you want to extend the OpenSim library by writing your own
+C++ plugin or you want to contribute directly to the OpenSim library, see the
+\a developer API documentation.
```
Comment (sherm1) [inline OpenSim/OpenSimDoxygenMain.h]: Now that we support Markdown in doxygen, you can just write \`developer\` in back quotes which I think is a much nicer syntax that doxygen's `\a developer`. It also works for multiple word phrases.

**Comment by sherm1** [comment]
Done reviewing -- I love it! Great job, Chris.

**Comment by chrisdembia** [comment]
I rebased and had to force-push. I hope no one else used this branch before the rebase.

@sherm1 I addressed all your comments.

**Comment by sherm1** [comment]
I think this is ready to merge but probably @jimmyDunne should at least bless it first.

**Comment by jenhicks** [comment]
Thanks @chrisdembia ! I have a couple comments and questions:

1) I think it would be helpful if the pages looked more visually different. Is it possible to  make the "OpenSim 3.2 Documentation" title (black/blue text) contain the distinguishing info between the two versions. That the part that I see first (rather than the header portion at the top). Could we even make the pages have different color schemes? 

2) Should existing links continue to point to the user version or developer version? There are arguments for both. User is probably what we want most people to see. Developer is a better match to what is there now. We should definitely add crosslinks between the two. 

3) Right now the google search box on the opensim support page searches the doxygen. Should it search both versions? Just user? We might be able to tell it to show results for one first, but I'm not 100% sure. 

4) I assume that as we continue to update the doxygen, these two version can diverge further (e.g., separate info using /if user and /if developer).    

5) And a side note - I noticed that the list of links to "Simbody classes of interest" links to your local version of Simbody, Chris (per issue #122). The link in the image of the API organization still has a hard coded link to the online version. What happens when we put the page online?

**Comment by chrisdembia** [comment]
Thanks @jenhicks.

> "OpenSim 3.2 Documentation" title (black/blue text) contain the distinguishing info between the two versions. That the part that I see first (rather than the header portion at the top). Could we even make the pages have different color schemes?

I see that title first, as well. I think that will be possible. As for the colors, that's a great idea and is extremely easy to change.


> Should existing links continue to point to the user version or developer version?

I vote for the links to the point to the user doxygen. Alternatively, we could have an extremely simple "landing documentation page" [like SWIG does](http://www.swig.org/Doc3.0/index.html) that just contains two links, for the user and dev pages.

> Should it search both versions? Just user?

Just user I think. The developer doxygen is for someone who is implementing a specific class, so they will already know which page they want to go to. If you already know which class you want, you don't need the google search. I should clarify this in my description of the developer doxygen.

> 4) I assume that as we continue to update the doxygen, these two version can diverge further (e.g., separate info using /if user and /if developer).

Yeah...I haven't thought about what the effects of this might be. Could be bad, could be good. I expect that people in general will not remember to use the `\if user` / `\if developer` logic. We can change this aspect as the documentation evolves.

> 5) And a side note - I noticed that the list of links to "Simbody classes of interest" links to your local version of Simbody, Chris (per issue #122). The link in the image of the API organization still has a hard coded link to the online version. What happens when we put the page online?

When we properly put the pages online, those links will point to Simbody's online documentation. Read [this](https://github.com/opensim-org/opensim-core/blob/master/OpenSim/doc/CMakeLists.txt#L12) and please let me know if it is unclear. The hardlinks in that diagram (#122) is still something I'd like to fix, but I really don't know how we do it.

Thanks for the review.

**Comment by chrisdembia** [comment]
Would anyone like any changes to this PR or is this ready to merge?

**Comment by jenhicks** [comment]
Did you update the color/title per my comment? I would keep the user one blue (the current color scheme) and make the developer one a different color.

**Comment by chrisdembia** [comment]
> Did you update the color/title per my comment? I would keep the user one blue (the current color scheme) and make the developer one a different color.

I realized last week that I forgot to update the colors; sorry about that. I like your suggestion. I thought it would be really easy to do this, but since we are using a custom theme it'll be harder. 

As for the title, I think I made that change but I don't remember. I'll post to this PR again when I'm sure both issuses are addressed.

**Comment by chrisdembia** [comment]
Hey @jenhicks, I have updated the colors and titles.

View both doxygen outputs here: http://stanford.edu/~dembia/opensim-user-vs-dev/

User:
image

Developer:
image


I chose a more descriptive title to replace "OpenSim 3.2 documentation": "Overview", and I put the user-vs-dev title in the banner up top. @jimmyDunne made a good point; the banner is on every page, while "Overview" only appears on the main page.

**Comment by aseth1** [comment]
The color differences and gradients did not appear for me on Firefox (39.0)

**Comment by chrisdembia** [comment]
I'll remove the `\if user` and `\if developer`, and we can rely only on `\internal.

**Comment by chrisdembia** [comment]
@aseth1 I tried to get distinct colors in Firefox. Are these two banners distinct enough?

user:
image

developer:
image

As for `\if user`, `\if developer`: I left these in, as I realize I use them to tailor the front page documentation. However, I removed the documentation in CONTRIBUTING.md about them, instead just telling people to use `\internal`. So the `\if user/developer` is now like a hidden feature (but I needed it to tailer the front page).

**Comment by chrisdembia** [comment]
@jenhicks how about these colors?
user:
image

developer:
image

I searched online for a while about resources for selecting colors that show well on all browsers, but did not find anything good. I found some stuff on "color management" but that seemed like overkill, and to be more about monitors.

**Comment by chrisdembia** [comment]
By the way I did not commit the updated colors yet; they're just on my computer.

**Comment by jenhicks** [comment]
Looks good to me! (I think I was thinking of photoshop.)

**Comment by chrisdembia** [comment]
Would either @jenhicks or @aseth1 like to merge this?

**Comment by chrisdembia** [comment]
I just updated this PR to remove a merge conflict. Would someone like to merge this?

**Comment by chrisdembia** [comment]
Thank you Jen.

**Pull request merged**